### DISCUSSION TITLE
Unable to build build-scripts, fails with permission denied

### DISCUSSION THREAD

**Discussion by TheNeikos** [ipetkov/crane#203] [Q&A] [answered]
Hello!

I'm currently having problems building my crate with crane, as trying to do so errors on a build script:

```
++ command cargo build --profile release --message-format json-render-diagnostics --all-features --all
   Compiling libc v0.2.139
   Compiling version_check v0.9.4
   Compiling log v0.4.17
   Compiling memchr v2.5.0
   Compiling cfg-if v1.0.0
   Compiling bitflags v1.3.2
   Compiling io-lifetimes v1.0.3
error: failed to run custom build command for `memchr v2.5.0`

Caused by:
  could not execute process `/build/hzaiznmrgahci7xfghnndal7anrmcpwj-source/target/release/build/memchr-639809555035ba6c/build-script>

Caused by:
  Permission denied (os error 13)
```

To reproduce, one can run this command:
`nix flake check github:TheNeikos/cloudmqtt/feature/add_mqtt_tester_to_flake`

It works fine in the development shell of that flake. 

I've tried looking for a reason for this everywhere, but have not found anything :/ 

Thanks for insights!

**Comment by ipetkov** [comment, marked as answer]
Thanks for the report @TheNeikos !

One thing I noticed about your flake is that the `buildDepsOnly` closure is not building the entire workspace, while the clippy derivation builds all crates with all features. Although you likely don't want to be doing this (for caching efficiency, changing the flags will require cargo to rebuild parts of the dependency closure which won't get cached) it seems to be tickling the actual bug. Here's what I think is happening

1. deps-only closure just builds the root crate, which likely depends on `memchr` somewhere
2. this derivation builds `memchr'`s build script, which is writable and executable, and runs it with no problem
3. the subsequent derivations start building the entire workspace with all features. some other feature or dependency eventually leads to `memchr` needing to be rebuilt (possibly due to its own feature flag changing, or that of a dependency it has)
4. cargo sees that the build script is already compiled from the inherited artifacts and tries to execute it.
5. the artifact inheritance ends up losing the executable bit on the script, which fails with the permissions error above

```nix
# Transcribed for posterity
let
  cargoArtifacts = craneLib.buildDepsOnly {
    inherit src;
  };

  cloudmqtt = craneLib.buildPackage {
    inherit cargoArtifacts src version;
    cargoExtraArgs = "--all-features --all";
  };

in
{
  checks = {
    inherit cloudmqtt;

    cloudmqtt-clippy = craneLib.cargoClippy {
      inherit cargoArtifacts src;
      cargoExtraArgs = "--all --all-features";
      cargoClippyExtraArgs = "-- --deny warnings";
    };
  };
}
```

**Reply by ipetkov** [reply]
This should be fixed with https://github.com/ipetkov/crane/pull/206 !

**Reply by TheNeikos** [reply]
A-mazing that you managed to figure that out! Thank you so much for maintaining all this and finding/fixing the bug! <3

**Reply by ipetkov** [reply]
Thanks for the reproduction flake, made it easy to track down the bug and write a minimal test case too :smile:
